### PR TITLE
doc: add interrupt implementation details

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Issue: ZEP-634
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by nashif — APPROVED** [review]

File: doc/kernel/other/interrupts.rst [hunk 1]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+    };
+
+This is then used by the gen_isr_tables.py script to
+generate a vector table and/or a software ISR table. This is produced as a C
+file and linked into the final application.
+
+The priority level of any interrupt is not encoded in these tables, instead
+:c:macro:`IRQ_CONNECT` also has a runtime component which programs the
+desired priority level of the interrupt to the interrupt controller. Some
+arches do not support the notion of interrupt priority, in which case the
```
Comment (nashif) [inline doc/kernel/other/interrupts.rst]: architectures

File: doc/kernel/other/interrupts.rst [hunk 2]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+priority argument is ignored.
+
+Vector Table
+------------
+Whether to generate this depends on whether CONFIG_GEN_IRQ_VECTOR_TABLE is
```
Comment (nashif) [inline doc/kernel/other/interrupts.rst]: Whether to generate the vector table ...

File: doc/kernel/other/interrupts.rst [hunk 3]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+    };
+
+This is used by the common software IRQ handler to look up the ISR and its
+argument and execute it. The active IRQ line is looked up in an interrupt
+controller register and used to index this table.
+
+x86 Details
+-----------
+
+x86 has a special type of vector table called the Interrupt Descriptor Table
```
Comment (nashif) [inline doc/kernel/other/interrupts.rst]: The X86 architecture has a special ....

**Review by nashif — CHANGES_REQUESTED** [review]

File: doc/kernel/other/interrupts.rst [hunk 4]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
        ...
     }
 
+Implementation Details
+======================
+
+Interrupt tables are set up at build time using some special build tools.
+The details laid out here apply to all architectures except x86, which will
+be covered in its own section.
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: how about:
```
    which are covered in the `x86 Details`_ section below.
```

File: doc/kernel/other/interrupts.rst [hunk 5]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+
+    struct {
+        void *spurious_irq_handler;
+        void *sw_irq_handler;
+        u32_t num_isrs;
+        u32_t num_vectors;
+        struct _isr_list isrs[];  <- of size num_isrs
+    };
+
+This is then used by the gen_isr_tables.py script to
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: Say what "this" is.  Are you referring to zephyr_prebuild.elf?

File: doc/kernel/other/interrupts.rst [hunk 6]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+generate a vector table and/or a software ISR table. This is produced as a C
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: generate a C file defining a vector table and software ISR table that are then compiled and linked into the final application.

File: doc/kernel/other/interrupts.rst (hunk 2 above)
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: How about:
```
A vector table is generated when CONFIG_GEN_IRQ_VECTOR_TABLE is enabled.
```

File: doc/kernel/other/interrupts.rst [hunk 7]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+The priority level of any interrupt is not encoded in these tables, instead
+:c:macro:`IRQ_CONNECT` also has a runtime component which programs the
+desired priority level of the interrupt to the interrupt controller. Some
+arches do not support the notion of interrupt priority, in which case the
+priority argument is ignored.
+
+Vector Table
+------------
+Whether to generate this depends on whether CONFIG_GEN_IRQ_VECTOR_TABLE is
+enabled. This is simply an array of function pointers, where each element
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: A vector table is simply an array of function pointers, where each element **n** corresponds to the IRQ handler for IRQ line **n**, and the function pointers are:

File: doc/kernel/other/interrupts.rst [hunk 8]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+#. For 'direct' interrupts declared with :c:macro:`IRQ_DIRECT_CONNECT`, the
+   handler function will be placed here.
+#. For regular interrupts declared with :c:macro:`IRQ_CONNECT`, the address
+   of the common software IRQ handler is placed here. This code does common
+   kernel interrupt bookkeeping and looks up the ISR and parameter from the
+   software ISR table.
+#. For interrupt lines that are not configured at all, the address of the
+   spurious IRQ handler, which causes a system fatal error if encountered.
+
+Some architectures (like Nios II internal interrupt controller) have a common
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: s/like/such as/

File: doc/kernel/other/interrupts.rst [hunk 9]
```diff
@@ -206,6 +206,128 @@ The following code demonstrates a direct ISR:
...
+entry point for all interrupts and do not support a vector table, in which case
+this option should be disabled.
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: the CONFIG_GEN_IRQ_VECTOR_TABLE option should be disabled.

**Review by dbkinder — CHANGES_REQUESTED** [review]
some clarification changes

**Review by nashif — DISMISSED** [review]

File: doc/kernel/other/interrupts.rst [hunk 10]
```diff
@@ -206,6 +206,129 @@ The following code demonstrates a direct ISR:
        ...
     }
 
+Implementation Details
+======================
+
+Interrupt tables are set up at build time using some special build tools.  The
+details laid out here apply to all architectures except x86, which which are
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: delete the extra "which"

File: doc/kernel/other/interrupts.rst [hunk 11]
```diff
@@ -206,6 +206,129 @@ The following code demonstrates a direct ISR:
...
+and the function pointers are:
+
+#. For 'direct' interrupts declared with :c:macro:`IRQ_DIRECT_CONNECT`, the
+   handler function will be placed here.
+#. For regular interrupts declared with :c:macro:`IRQ_CONNECT`, the address
+   of the common software IRQ handler is placed here. This code does common
+   kernel interrupt bookkeeping and looks up the ISR and parameter from the
+   software ISR table.
+#. For interrupt lines that are not configured at all, the address of the
+   spurious IRQ handler, which causes a system fatal error if encountered.
```
Comment (dbkinder) [inline doc/kernel/other/interrupts.rst]: after "which causes a system fatal error if encoungered" add ", will be placed here."

**Review by dbkinder — CHANGES_REQUESTED** [review]
a few (final) suggestions

**Comment by andrewboie** [comment]
updated

**Pull request merged**